### DISCUSSION TITLE
V7 - Typing of useFieldArray

### DISCUSSION THREAD

**Discussion by guyzmo** [react-hook-form/react-hook-form#5026] [Q&A] [answered]
Hello,

I'm trying to convert an app using RHFv6 to v7, and I'm struggling with converting the field arrays.

My forms are splits across many components, thus I often have a pattern with the `useForm` provided as context to other components in an outter component.

Then I have the inner component with a field array, and I struggle to get proper types for it:

I first try the naive approach, I don't give it types or controller (I did not expect that to work):

```
const myArray = useFieldArray({
  name: "ArrayFieldName",
  keyName: "key" // mandatory as my data is from graphql and default id provoques a name clash
})
```

Then the type of `myArray.fields` is `Record<"key", string>[]`.

```
const form = useFormContext<MyFormType>();
const myArray = useFieldArray({
  control: form.control,
  name: "ArrayFieldName",
  keyName: "key"
})
```

Then the type of `myArray.fields` is `never[]`.

If I try telling it what my types are:

```
const myArray = useFieldArray<MyArrayFieldType, "key">({
  name: "ArrayFieldName",
  keyName: "key"
})
```

I get the following error on `"key"` generic parameter: `Type 'string' does not satisfy the constraint 'never'.`

I then I tried digging into the source, and found out key is the third argument (it's not in the V7 documentation), so I tried several things, but the second generic argument only accepts `never`.

So there's something I really do not understand here on how to properly type `useFieldArray`s in V7, and it might be a good idea to update the documentation (I can make a ticket about that after the discussion).

BTW, those two pages:

* https://react-hook-form.com/ts#UseFieldArrayMethods
* https://react-hook-form.com/ts#UseFieldArrayOptions

are in the V7 documentation, and yet are deprecated (the names have changed to Return/Props). It's likely to have fall through from V6 documentation, and it'd be good to change it, or at least make an annotation to tell it will change but it's still V6 material.

Thank you for the hard work :+1:

**Comment by bluebill1049** [comment]
Hey @guyzmo sorry to hear the frasutration. Do you want to take a look at this CodeSandbox? 

https://codesandbox.io/s/calc-i231d

I think what you are missing is pass generic to `control`.

**Comment by thanh-nguyen-95** [comment, marked as answer]
Here's how I do it in my project

```ts
type Bar = {
  id: number
  someKey: string
}

interface FormValues {
  foo: string
  bars: Bar[]
}

// root (parent component)
const formMethods = useForm<FormValues>()

// inside components
const { control } = useFormContext<FormValues>()
const { fields } = useFieldArray({ name: 'bars', keyName: '_id', control })
```

The type of each `field` will be inferred as `Bar & { _id: string }`. It's important to pass in `control` else it doesn't work.

Hope this helps, cheers!

**Reply by bluebill1049** [reply]
Thanks for sharing solution.

**Comment by grnwld** [comment]
HI @guyzmo, did you work this out? I see you're using generic when calling `useFormContext`, as am I, and getting the same results. Any help would be appreciated; thanks! :)

**Reply by grnwld** [reply]
Just to follow-up, it's to do with optional fields. I have a type with the optional field `{dependencies?: Dependency[]}`. Changing it to non-optional fixed the problem. Not sure if this is by design or not, so mentioning at least for anyone else who runs into the problem.

**Reply by guyzmo** [reply]
Hello, I did not have time to try again to port the form to v7, I'm still on v6.
I'm actually having double the work as I'm taking on work for a colleague who's been sick for a few months ☹

I'll keep you updated as soon as I try again to upgrade ☺
